sky/task.py

```diff
@@ -68,6 +69,35 @@ def _is_valid_env_var(name: str) -> bool:
     return bool(re.fullmatch(_VALID_ENV_VAR_REGEX, name))
 
 
+def _fill_in_env_vars_in_file_mounts(
+    file_mounts: Dict[str, Any],
+    task_envs: Dict[str, str],
+) -> Dict[str, Any]:
+    """Detects env vars in file_mounts and fills them with task_envs.
+
+    Use cases of env vars in file_mounts:
+    - dst/src paths; e.g.,
+        /model_path/llama-${SIZE}b: s3://llama-weights/llama-${SIZE}b
+    - storage's name (bucket name)
+    - storage's source (local path)
+
+    We simply dump file_mounts into a json string, and replace env vars using
+    regex. This should be safe as file_mounts has been schema-validated.
+
+    Env vars of the following forms are detected:
+        - ${ENV}
+        - $ENV
+    where <ENV> must appear in task.envs.
+    """
+    # TODO(zongheng): support ${ENV:-default}?
+    file_mounts_str = json.dumps(file_mounts)
+    for key, value in task_envs.items():
+        pattern = r'\$\{?\b' + key + r'\b\}?'
```

Comment:
nit: it seems bash replaces the envvar by looking for the envvar in the string first and match it with the existing envvar, while here we do it in a reversed way. That will cause a slight inconsistency in the behavior:
```
HELLO_WORLD=1
echo $HELLO_WORLD_1
```
The block above will output an empty string, but in our implementation, we will output `1_1`.

Probably, an alternative way to do this is:
```
# Create a replacement function
def replace_var(match):
    var_name = match.group(1)  # Now directly accessing the variable name
    return task_envs.get(var_name, match.group(0))  # If the variable isn't in the dictionary, return it unchanged

pattern = r'\$\{?\b([a-zA-Z_][a-zA-Z0-9_]*)\b\}?'
# Use re.sub with the replacement function
result = re.sub(pattern, replace_var, file_mounts_str)
```

This is a very corner case, we can also leave it as is to see how people feel about it.

Reply:
Great catch! Used this & reran the smoke test.